thunder/executors/functional_teex.py

```diff
@@ -0,0 +1,236 @@
+from transformer_engine.pytorch.tensor.float8_tensor import Float8Quantizer
+from transformer_engine.pytorch.ops import BasicLinear
+from transformer_engine.pytorch.fp8 import (
+    _amax_and_scale_update,
+    get_fp8_max,
+    get_default_fp8_recipe,
+    Recipe,
+    RecipeState,
+)
+
+
+class StatefulExecutor(OperatorExecutor):
+    def __init__(self, name, *, version=None):
+        super().__init__(name, version=version)
+
+    def register_stateful_operator(self, base_name: str, state_class, *, meta):
+        def register_state(*args, **kwargs):
+            state = state_class()
+            name = f"{base_name}_{id(state)}"
+
+            def bind_state(bsym):
+                bsym._call_ctx = {name: state}
+
+            sym = self.register_operator(name, meta=meta, bind_postprocess=bind_state)
+            return sym(*args, *kwargs)
+
+        return register_state
+
+
+functional_te_ex = StatefulExecutor("functional_te")
+register_executor(functional_te_ex)
+
+
+# TODO
+def _functional_te_checker(a, w, /, bias):
+    # TE 2.0 functional API does not yet support bias=True
+    # https://github.com/NVIDIA/TransformerEngine/blob/1321b9b5dc96d67d20d6682c52116a3657f293d3/transformer_engine/pytorch/ops/basic/basic_linear.py#L46-L47
+    # https://github.com/NVIDIA/TransformerEngine/blob/1321b9b5dc96d67d20d6682c52116a3657f293d3/transformer_engine/pytorch/ops/basic/basic_linear.py#L505
+    if bias is not None:
+        return False
+    return _linear_checker(a, w, bias)
+
+
+# Assuming one recipe per machine since it's based on hw capabilities.
+functional_te_recipe = get_default_fp8_recipe()
+
+
+def _te_fp8_recipe_meta(recipe_name: str):
+    return AnyProxy(recipe_name, prefix="r")
+
+
+def _te_fp8_recipe_impl(recipe_name: str):
+    return functional_te_recipe
+
+
+_te_fp8_recipe = functional_te_ex.register_operator("te_fp8_recipe", meta=_te_fp8_recipe_meta, fn=_te_fp8_recipe_impl)
+
+
+def _te_fp8_quantizers_meta(recipe_state: RecipeState, num_quantizers: int):
+    # giving num quantizers so there is no need to allocate extra things
+    return (*(AnyProxy(num_quantizers, prefix="q") for _ in range(num_quantizers)),)
+
+
+class TEQuantizerState:
+    def __init__(self):
+        self.quantizers = None
+
+    def __call__(self, recipe_state: RecipeState, num_quantizers: int):
+        if self.quantizers:
+            return self.quantizers
+        quantizers = recipe_state.make_quantizers()
+
+        self.quantizers = quantizers
+
+        return quantizers
+
+
+_te_fp8_quantizers = functional_te_ex.register_stateful_operator(
+    "te_fp8_quantizers", TEQuantizerState, meta=_te_fp8_quantizers_meta
+)
+
+
+def _te_fp8_state_meta(recipe, mode: str, num_quantizers: int, /):
+    # mode just to give the proxy something wihout allocating extra stuff
+    return AnyProxy(mode, prefix="s")
+
+
+class TERecipeState:
+    def __init__(self):
+        self.state = None
+
+    def __call__(self, recipe: Recipe, mode, num_quantizers):
```

Comment:
It'd be nice to mention what `mode` is as a comment. 

Had to look at TE to understand: https://github.com/NVIDIA/TransformerEngine/blob/1321b9b5dc96d67d20d6682c52116a3657f293d3/transformer_engine/pytorch/fp8.py#L913-L921